### ISSUE TITLE
About space complexity

### ISSUE THREAD

**Issue by llan-ml** [lucfra/LDS-GNN#6]
Hi @lucfra 

I tried to run LDS on the dataset Pubmed, which has 19,717 nodes and 44,338 edges. The dense adjacent matrix should occupy about `19717 * 19717 * 32 / 8 / 10e9 = 1.56 GB` memory.

When we directly train a GCN model on Pubmed, the process occupies about `684 MB` GPU memory (since the adjacent matrix is sparse).

On the other hand, when we train a GCN model in an LDS manner, the process occupies `30.95 GB` GPU memory. 

Could you provide some space complexity analysis of LDS? I did not found it in the paper. Thanks!

**Comment by lucfra** [comment]
Hi @llan-ml 

Informally, the most expensive operations in term of memory requirement should be those for computing d_\theta w_T with reverse mode differentiation and should take up to 5 times the memory required by the learning dynamics (by default Adam, which uses 3 times the memory of GD). Since the biggest object is the adjacency matrix, a very rough calculation gives
1.56      *          3       *       5 =  23.4 Gb 
(adj. mat)     (adam)    (RMD)

which is not too far from the actual consumption. One quick way around could be to use gradient descent instead of adam, but of course this does not really solve the issue of scalability.

It would be interesting to develop a stochastic version of the algorithm sampling a subset of nodes at each inner iteration, but unfortunately this is not straightforward, and we are not planning to do it in the immediate future.  

Hope this helps, 

Cheers, 
Luca

**Comment by llan-ml** [comment]
I got it. Thank you very much for your reply.
